File: projectforge-business/src/main/java/org/projectforge/framework/persistence/api/QueryFilter.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2019 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.framework.persistence.api;

import org.apache.commons.lang3.StringUtils;
import org.hibernate.Criteria;
import org.hibernate.FetchMode;
import org.hibernate.Session;
import org.hibernate.criterion.Criterion;
import org.hibernate.criterion.Order;
import org.hibernate.criterion.Restrictions;
import org.projectforge.business.multitenancy.TenantService;
import org.projectforge.framework.configuration.ApplicationContextProvider;
import org.projectforge.framework.persistence.user.api.ThreadLocalUserContext;
import org.projectforge.framework.persistence.user.api.UserContext;
import org.projectforge.framework.persistence.user.entities.TenantDO;
import org.projectforge.framework.time.DateHelper;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

/**
 * Stores the expressions and settings for creating a hibernate criteria object. This template is useful for avoiding
 * the need of a hibernate session in the stripes action classes.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
public class QueryFilter {
  private final List<Object> filterSettings = new ArrayList<Object>();

  private String name;

  private String alias;

  private BaseSearchFilter filter;

  private FetchMode fetchMode;

  private String associationPath = null;

  private Locale locale;

  /**
   * Creates new QueryFilter with a new SearchFilter as filter.
   */
  public QueryFilter() {
    this.filter = new BaseSearchFilter();
  }

  public QueryFilter(final BaseSearchFilter filter) {
    this(filter, false);
  }

  /**
   * @param filter
   * @param ignoreTenant default is false.
   */
  public QueryFilter(final BaseSearchFilter filter, final boolean ignoreTenant) {
    if (filter == null) {
      this.filter = new BaseSearchFilter();
    } else {
      this.filter = filter;
    }
    TenantService tenantService = ApplicationContextProvider.getApplicationContext().getBean(TenantService.class);
    if (ignoreTenant == false && tenantService.isMultiTenancyAvailable() == true) {
      final UserContext userContext = ThreadLocalUserContext.getUserContext();
      final TenantDO currentTenant = userContext.getCurrentTenant();
      if (currentTenant != null) {
        if (currentTenant.isDefault() == true) {
          this.add(Restrictions.or(Restrictions.eq("tenant", userContext.getCurrentTenant()),
                  Restrictions.isNull("tenant")));
        } else {
          this.add(Restrictions.eq("tenant", userContext.getCurrentTenant()));
        }
      }
    }
  }

  private QueryFilter(final String name) {
    this.name = name;
  }

  public String getName() {
    return name;
  }

  public String getAlias() {
    return alias;
  }

  public BaseSearchFilter getFilter() {
    return filter;
  }

  /**
   * Locale is needed for lucene stemmers (hibernate search).
   *
   * @return
   */
  public Locale getLocale() {
    if (locale == null) {
      return Locale.GERMAN;
    }
    return locale;
  }

  public void setLocale(final Locale locale) {
    this.locale = locale;
  }

  /**
   * If an error occured (e. g. lucene parse exception) this message will be returned.
   *
   * @return
   */
  public String getErrorMessage() {
    return filter.getErrorMessage();
  }

  public void setErrorMessage(final String errorMessage) {
    filter.setErrorMessage(errorMessage);
  }

  public boolean hasErrorMessage() {
    return filter.hasErrorMessage();
  }

  public void clearErrorMessage() {
    filter.clearErrorMessage();
  }

  /**
   * @param criterion
   * @return
   * @see org.hibernate.Criteria#add(Criterion)
   */
  public QueryFilter add(final Criterion criterion) {
    filterSettings.add(criterion);
    return this;
  }

  /**
   * @param order
   * @return
   * @see org.hibernate.Criteria#addOrder(Order)
   */
  public QueryFilter addOrder(final Order order) {
    filterSettings.add(order);
    return this;
  }

  public void setFetchMode(final String associationPath, final FetchMode mode) {
    this.associationPath = associationPath;
    this.fetchMode = mode;
  }

  /**
   * Adds Expression.between for given time period.
   *
   * @param dateField
   * @param year      if <= 0 do nothing.
   * @param month     if < 0 choose whole year, otherwise given month. (Calendar.MONTH);
   */
  public void setYearAndMonth(final String dateField, final int year, final int month) {
    if (year > 0) {
      final Calendar cal = DateHelper.getUTCCalendar();
      cal.set(Calendar.YEAR, year);
      java.sql.Date lo = null;
      java.sql.Date hi = null;
      if (month >= 0) {
        cal.set(Calendar.MONTH, month);
        cal.set(Calendar.DAY_OF_MONTH, 1);
        lo = new java.sql.Date(cal.getTimeInMillis());
        final int lastDayOfMonth = cal.getActualMaximum(Calendar.DAY_OF_MONTH);
        cal.set(Calendar.DAY_OF_MONTH, lastDayOfMonth);
        hi = new java.sql.Date(cal.getTimeInMillis());
      } else {
        cal.set(Calendar.DAY_OF_YEAR, 1);
        lo = new java.sql.Date(cal.getTimeInMillis());
        final int lastDayOfYear = cal.getActualMaximum(Calendar.DAY_OF_YEAR);
        cal.set(Calendar.DAY_OF_YEAR, lastDayOfYear);
        hi = new java.sql.Date(cal.getTimeInMillis());
      }
      add(Restrictions.between(dateField, lo, hi));
    }
  }

  public Criteria buildCriteria(final Session session, final Class<?> clazz) {
    final Criteria criteria = session.createCriteria(clazz);
    buildCriteria(criteria);
    return criteria;
  }

  private void buildCriteria(final Criteria criteria) {
    for (final Object obj : filterSettings) {
      if (obj instanceof Criterion) {
        criteria.add((Criterion) obj);
      } else if (obj instanceof Order) {
        criteria.addOrder((Order) obj);
      } else if (obj instanceof Alias) {
        final Alias alias = (Alias) obj;
        criteria.createAlias(alias.arg0, alias.arg1, alias.joinType);
      } else if (obj instanceof QueryFilter) {
        final QueryFilter filter = (QueryFilter) obj;
        Criteria subCriteria;
        if (StringUtils.isEmpty(filter.getAlias()) == true) {
          subCriteria = criteria.createCriteria(filter.getName());
        } else {
          subCriteria = criteria.createCriteria(filter.getName(), filter.getAlias());
        }
        filter.buildCriteria(subCriteria);
      }
    }
    if (associationPath != null) {
      criteria.setFetchMode(associationPath, fetchMode);
    }
    if (filter.sortAndLimitMaxRowsWhileSelect) {
      if (filter.getMaxRows() > 0) {
        criteria.setMaxResults(filter.getMaxRows());
      }
      if (filter.getSortProperty() != null) {
        if (filter.getSortOrder() == SortOrder.DESCENDING)
          criteria.addOrder(Order.desc(filter.getSortProperty()));
        else
          criteria.addOrder(Order.asc(filter.getSortProperty()));
      }
    }
  }

  /**
   * @see org.hibernate.Criteria#createAlias(String, String)
   */
  public QueryFilter createAlias(final String arg0, final String arg1) {
    filterSettings.add(new Alias(arg0, arg1));
    return this;
  }

  /**
   * @see org.hibernate.Criteria#createAlias(String, String, int)
   */
  public QueryFilter createAlias(final String arg0, final String arg1, final int joinType) {
    filterSettings.add(new Alias(arg0, arg1, joinType));
    return this;
  }

  public QueryFilter createCriteria(final String name) {
    final QueryFilter filter = new QueryFilter(name);
    filterSettings.add(filter);
    return filter;
  }

  public QueryFilter createCriteria(final String name, final String alias) {
    final QueryFilter filter = new QueryFilter(name);
    filter.alias = alias;
    filterSettings.add(filter);
    return filter;
  }

  class Alias {
    String arg0;

    String arg1;

    int joinType = Criteria.INNER_JOIN;

    Alias(final String arg0, final String arg1) {
      this.arg0 = arg0;
      this.arg1 = arg1;
    }

    Alias(final String arg0, final String arg1, final int joinType) {
      this.arg0 = arg0;
      this.arg1 = arg1;
      this.joinType = joinType;
    }
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/framework/persistence/api/QueryFilter.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2020 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.framework.persistence.api

import org.projectforge.business.tasktree.TaskTreeHelper
import org.projectforge.framework.ToStringUtil
import org.projectforge.framework.persistence.api.impl.DBFilter
import org.projectforge.framework.persistence.api.impl.DBHistorySearchParams
import org.projectforge.framework.persistence.api.impl.DBJoin
import org.projectforge.framework.persistence.api.impl.DBPredicate
import org.projectforge.framework.time.PFDateTime
import org.projectforge.framework.time.PFDay
import org.slf4j.LoggerFactory
import java.time.LocalDate
import java.time.Month
import javax.persistence.criteria.JoinType

/**
 * If no maximum number of results is defined, MAX_ROWS is used as max value.
 */
const val QUERY_FILTER_MAX_ROWS: Int = 10000

/**
 * Convenient helper to create database queries (criteria search, full text search and search in result lists).
 * It will be automatically detected, which kind of database query is needed (critery, full text or multi field full text query).
 * Field of the index of hibernate search will be detected and used in full text queries, if not indexed, the result will be filtered.
 *
 * You may add your predicates (independent of which strategy is used behind). The query strategy is automatically done.
 *
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
class QueryFilter @JvmOverloads constructor(filter: BaseSearchFilter? = null,
                                            val ignoreTenant: Boolean = false) {
    private val predicates = mutableListOf<DBPredicate>()

    val joinList = mutableListOf<DBJoin>()

    var sortProperties = mutableListOf<SortProperty>()

    var fullTextSearchFields: Array<String>? = null

    private val historyQuery = DBHistorySearchParams()

    /**
     * If true, any searchstring (alphanumeric) without wildcard will be changed to '<searchString>*'.
     */
    var autoWildcardSearch: Boolean = false

    /**
     * If null, deleted and normal entries will be queried.
     */
    var deleted: Boolean? = null

    /**
     * Extend the filter by additional variables and settings.
     */
    var extended: MutableMap<String, Any> = mutableMapOf()

    var searchHistory: String?
        get() = historyQuery.searchHistory
        set(value) {
            historyQuery.searchHistory = value
        }

    var modifiedFrom: PFDateTime?
        get() = historyQuery.modifiedFrom
        set(value) {
            historyQuery.modifiedFrom = value
        }

    var modifiedTo: PFDateTime?
        get() = historyQuery.modifiedTo
        set(value) {
            historyQuery.modifiedTo = value
        }

    var modifiedByUserId: Int?
        get() = historyQuery.modifiedByUserId
        set(value) {
            historyQuery.modifiedByUserId = value
        }

    var maxRows: Int = QUERY_FILTER_MAX_ROWS

    var sortAndLimitMaxRowsWhileSelect: Boolean = true

    fun getExtendedBooleanValue(key: String): Boolean {
        val value = extended[key] ?: return false
        if (value is Boolean) {
            return value
        }
        return false
    }

    init {
        if (filter != null) {
            this.fullTextSearchFields = filter.fullTextSearchFields
            this.autoWildcardSearch = true
            // Legacy for old implementation:
            if (!filter.ignoreDeleted) {
                deleted = filter.deleted
            }
            if (filter.isSearchHistory && !filter.searchString.isNullOrBlank()) {
                searchHistory = filter.searchString
            }
            if (filter.isSearchNotEmpty) {
                addFullTextSearch(filter.searchString, autoWildcardSearch)
            }
            if (filter.useModificationFilter) {
                if (filter.modifiedSince != null) {
                    modifiedFrom = PFDateTime.from(filter.modifiedSince) // not null
                } else if (filter.startTimeOfModification != null) { // not null
                    modifiedFrom = PFDateTime.from(filter.startTimeOfModification)
                }
                if (filter.stopTimeOfModification != null) {
                    modifiedTo = PFDateTime.from(filter.stopTimeOfModification) // not null
                }
            }
            if (filter.modifiedByUserId != null) {
                modifiedByUserId = filter.modifiedByUserId
            }
            if (filter.maxRows > 0) {
                maxRows = filter.maxRows
            }
        }
    }


    /**
     * @return this for chaining
     */
    fun add(predicate: DBPredicate): QueryFilter {
        predicates.add(predicate)
        return this
    }

    fun addOrder(vararg sortProperty: SortProperty): QueryFilter {
        sortProperties.addAll(sortProperty)
        return this
    }

    /**
     * Create an alias for criteria search, used for Joins.
     * @param attr The attribute to create a alias (JoinSet) for. Nested properties are supported (order.positions).
     * @param fetch
     * @param joinType [JoinType.INNER] is default.
     * @param parent If not given, root is used. If given, the parent is used as root path of the attr.
     * @return this for chaining.
     */
    @JvmOverloads
    fun createJoin(attr: String, joinType: JoinType = JoinType.INNER, fetch: Boolean = false, parent: String? = null): QueryFilter {
        joinList.add(DBJoin(attr, joinType, fetch, parent))
        return this
    }

    /**
     * Does nothing if str is null or blank.
     */
    fun addFullTextSearch(str: String?, autoWildcardSearch: Boolean = false) {
        if (str.isNullOrBlank()) return
        predicates.add(DBPredicate.FullSearch(str, autoWildcardSearch))
    }

    /**
     * Adds Expression.between for given time period.
     *
     * @param dateField
     * @param year      if <= 0 do nothing.
     * @param month     if <= 0 choose whole year, otherwise given month (1-January, ..., 12-December);
     */
    fun setYearAndMonth(dateField: String, year: Int, month: Int) {
        if (year > 0) {
            val lo: LocalDate
            val hi: LocalDate
            if (month > 0) {
                val date = PFDay.withDate(year, month, 1)
                lo = date.beginOfMonth.localDate
                hi = date.endOfMonth.localDate
            } else {
                val date = PFDay.withDate(year, Month.JANUARY, 1)
                lo = date.beginOfYear.localDate
                hi = date.endOfYear.localDate
            }
            add(between(dateField, lo, hi))
        }
    }

    fun createDBFilter(): DBFilter {
        val dbFilter = DBFilter(sortAndLimitMaxRowsWhileSelect, maxRows, fullTextSearchFields)
        if (predicates.none { it.field == "deleted" } && deleted != null) {
            // Adds deleted flag, if not already exist in predicates:
            dbFilter.predicates.add(DBPredicate.Equal("deleted", deleted == true))
        }
        predicates.forEach {
            dbFilter.predicates.add(it)
        }
        sortProperties.forEach {
            dbFilter.sortProperties.add(it)
        }
        return dbFilter
    }

    override fun toString(): String {
        return ToStringUtil.toJsonString(this)
    }

    companion object {
        private val log = LoggerFactory.getLogger(QueryFilter::class.java)

        @JvmStatic
        fun isNull(field: String): DBPredicate.IsNull {
            return DBPredicate.IsNull(field)
        }

        @JvmStatic
        fun isNotNull(field: String): DBPredicate.IsNotNull {
            return DBPredicate.IsNotNull(field)
        }

        @JvmStatic
        fun eq(field: String, value: Any): DBPredicate.Equal {
            return DBPredicate.Equal(field, value)
        }

        @JvmStatic
        fun ne(field: String, value: Any): DBPredicate.NotEqual {
            return DBPredicate.NotEqual(field, value)
        }

        @JvmOverloads
        @JvmStatic
        fun like(field: String, value: String, autoWildcardSearch: Boolean = false): DBPredicate.Like {
            return DBPredicate.Like(field, value, autoWildcardSearch)
        }

        @JvmStatic
        fun <O : Comparable<O>> between(field: String, from: O, to: O): DBPredicate.Between<O> {
            return DBPredicate.Between(field, from, to)
        }

        @JvmStatic
        fun <O : Comparable<O>> ge(field: String, value: O): DBPredicate.GreaterEqual<O> {
            return DBPredicate.GreaterEqual(field, value)
        }

        @JvmStatic
        fun <O : Comparable<O>> gt(field: String, value: O): DBPredicate.Greater<O> {
            return DBPredicate.Greater(field, value)
        }

        @JvmStatic
        fun <O : Comparable<O>> le(field: String, value: O): DBPredicate.LessEqual<O> {
            return DBPredicate.LessEqual(field, value)
        }

        @JvmStatic
        fun <O : Comparable<O>> lt(field: String, value: O): DBPredicate {
            return DBPredicate.Less(field, value)
        }

        /**
         * @param from if given, ge or between search is used.
         * @param to if given, le or between search is used.
         */
        @JvmStatic
        fun <T : Comparable<T>> interval(field: String, from: T?, to: T?): DBPredicate {
            if (from != null)
                return if (to != null)
                    DBPredicate.Between(field, from, to)
                else
                    DBPredicate.GreaterEqual(field, from)
            else if (to != null)
                return DBPredicate.LessEqual(field, to)
            throw UnsupportedOperationException("interval needs at least one value ('from' and/or 'to').")
        }

        @JvmStatic
        fun <T> isIn(field: String, values: Collection<*>): DBPredicate.IsIn<T> {
            @Suppress("UNCHECKED_CAST")
            return DBPredicate.IsIn(field, *(values.toTypedArray() as Array<T>))
        }

        @JvmStatic
        fun <T> isIn(field: String, vararg values: T): DBPredicate.IsIn<T> {
            return DBPredicate.IsIn(field, *values)
        }

        @JvmStatic
        fun not(matcher: DBPredicate): DBPredicate.Not {
            return DBPredicate.Not(matcher)
        }

        @JvmStatic
        fun and(vararg matchers: DBPredicate): DBPredicate.And {
            return DBPredicate.And(*matchers)
        }

        @JvmStatic
        fun or(vararg matchers: DBPredicate): DBPredicate.Or {
            return DBPredicate.Or(*matchers)
        }

        @JvmStatic
        fun taskSearch(field: String, taskId: Int?, recursive: Boolean): DBPredicate {
            if (taskId == null) {
                return DBPredicate.IsNull(field)
            }
            val node = TaskTreeHelper.getTaskTree().getTaskNodeById(taskId)
            return if (node == null) {
                log.warn("Can't query for given task id #$taskId, no such task node found.")
                DBPredicate.IsNull(field)
            } else {
                if (recursive) {
                    val taskIds = node.descendantIds
                    taskIds.add(node.id)
                    if (log.isDebugEnabled) {
                        log.debug("search in tasks: $taskIds")
                    }
                    DBPredicate.IsIn(field, taskIds)
                } else {
                    DBPredicate.Equal(field, taskId)
                }
            }
        }
    }
}
